src/util.ts

```diff
@@ -21,3 +23,68 @@ export function toArray<T>(iterator: Iterator<T>): T[] {
   }
   return array;
 }
+
+/**
+ * Constructs a new ts.CompilerHost that overlays sources in substituteSource
+ * over another ts.CompilerHost.
+ *
+ * @param substituteSource A map of source file name -> overlay source text.
+ */
+export function createSourceReplacingCompilerHost(
+    substituteSource: Map<string, string>, delegate: ts.CompilerHost): ts.CompilerHost {
+  return {
+    getSourceFile,
+    getCancellationToken: delegate.getCancellationToken,
+    getDefaultLibFileName: delegate.getDefaultLibFileName,
+    writeFile: delegate.writeFile,
+    getCurrentDirectory: delegate.getCurrentDirectory,
+    getCanonicalFileName: delegate.getCanonicalFileName,
+    useCaseSensitiveFileNames: delegate.useCaseSensitiveFileNames,
+    getNewLine: delegate.getNewLine,
+    fileExists: delegate.fileExists,
+    readFile: delegate.readFile,
+    directoryExists: delegate.directoryExists,
+    getDirectories: delegate.getDirectories,
+  };
+
+  function getSourceFile(
+      fileName: string, languageVersion: ts.ScriptTarget,
+      onError?: (message: string) => void): ts.SourceFile {
+    let path: string = ts.sys.resolvePath(fileName);
+    let sourceText = substituteSource.get(path);
+    if (sourceText) {
+      return ts.createSourceFile(fileName, sourceText, languageVersion);
+    }
+    return delegate.getSourceFile(path, languageVersion, onError);
+  }
+}
+
+/**
+ * Constructs a new ts.CompilerHost that overlays sources in substituteSource
+ * over another ts.CompilerHost.
+ *
+ * @param substituteSource A map of source file name -> overlay source text.
+ */
+export function createOutputRetainingCompilerHost(
+    outputFiles: Map<string, string>, delegate: ts.CompilerHost): ts.CompilerHost {
```

Comment:
Out of curiosity, why do this instead of passing the extra arg to program.emit()?  I think they have the extra param exactly for doing the sort of file-writing hook we do.